--- a/arch/powerpc/platforms/ps3/Kconfig
+++ b/arch/powerpc/platforms/ps3/Kconfig
@@ -148,4 +148,24 @@
 	  profiling support of the Cell processor with programs like
 	  oprofile and perfmon2, then say Y or M, otherwise say N.
 
+config PS3_NFLASH
+	tristate "PS3 NFLASH Storage Driver"
+	depends on PPC_PS3 && BLOCK
+	select PS3_STORAGE
+	help
+	  Include support for the PS3 NOR FLASH Storage.
+
+	  This support is required to access the PS3 nor flash disk.
+	  In general, all users will say Y or M.
+
+config PS3_VFLASH
+	tristate "PS3 VFLASH Storage Driver"
+	depends on PPC_PS3 && BLOCK && !PS3_FLASH
+	select PS3_STORAGE
+	help
+	  Include support for the PS3 Virtual FLASH Storage.
+
+	  This support is required to access the PS3 virtual flash disk.
+	  In general, all users will say Y or M.
+
 endmenu
