--- a/arch/powerpc/platforms/ps3/platform.h
+++ b/arch/powerpc/platforms/ps3/platform.h
@@ -86,7 +86,8 @@
 	PS3_DEV_TYPE_SB_USB = 4,
 	PS3_DEV_TYPE_STOR_ROM = TYPE_ROM,	/* 5 */
 	PS3_DEV_TYPE_SB_GPIO = 6,
-	PS3_DEV_TYPE_STOR_FLASH = TYPE_RBC,	/* 14 */
+	PS3_DEV_TYPE_STOR_VFLASH = TYPE_RBC,	/* 14 */
+	PS3_DEV_TYPE_STOR_FLASH = 254,
 };
 
 int ps3_repository_read_bus_str(unsigned int bus_index, const char *bus_str,
